src/booking/booking.service.ts

```diff
@@ -1,26 +1,81 @@
-import { Injectable } from '@nestjs/common';
+import {
+  Injectable,
+  NotFoundException,
+  UnprocessableEntityException,
+  BadRequestException,
+} from '@nestjs/common';
+import { PrismaService } from 'src/prisma/prisma.service';
 import { CreateBookingDto } from './dto/create-booking.dto';
 import { UpdateBookingDto } from './dto/update-booking.dto';
 
 @Injectable()
 export class BookingService {
-  create(createBookingDto: CreateBookingDto) {
-    return 'This action adds a new booking';
+  constructor(private readonly prisma: PrismaService) {}
+
+  async create(createBookingDto: CreateBookingDto) {
+    await this.validateCreateBooking(
+      createBookingDto.startDate,
+      createBookingDto.endDate,
+      createBookingDto.roomId,
+    );
+    return this.prisma.booking.create({ data: createBookingDto });
   }
 
   findAll() {
-    return `This action returns all booking`;
+    return this.prisma.booking.findMany();
   }
 
   findOne(id: number) {
-    return `This action returns a #${id} booking`;
+    return this.prisma.booking.findUnique({ where: { id } });
   }
 
-  update(id: number, updateBookingDto: UpdateBookingDto) {
-    return `This action updates a #${id} booking`;
+  async update(id: number, updateBookingDto: UpdateBookingDto) {
+    try {
+      const booking = await this.prisma.booking.findUnique({ where: { id } });
+      if (!booking) {
+        throw new NotFoundException(`Booking with ID ${id} not found`);
+      }
+
+      return this.prisma.booking.update({
+        where: { id },
+        data: updateBookingDto,
+      });
+    } catch (error) {
+      throw new BadRequestException(error.message);
+    }
+  }
+
+  async remove(id: number) {
+    try {
+      const booking = await this.prisma.booking.findUnique({ where: { id } });
+      if (!booking) {
+        throw new NotFoundException(`Booking with ID ${id} not found`);
+      }
+
+      await this.prisma.booking.delete({ where: { id } });
```

Comment:
foreign key constraint . Introduce a new field named isDeleted, set to true for soft deletion.Only completed and cancelled bookings can undergo soft deletion.